server/src/main/java/io/druid/segment/realtime/firehose/ReplayableFirehoseFactory.java

```diff
@@ -0,0 +1,313 @@
+  {
+    private final JsonFactory factory = new DefaultObjectMapper(
+        new SmileFactory().enable(SmileGenerator.Feature.CHECK_SHARED_STRING_VALUES)
+    ).getFactory();
+
+    private final List<File> files = Lists.newArrayList();
+    private final List<String> dimensions;
+
+    private int fileIndex = 0;
+    private JsonParser jsonParser;
+    private Iterator<Row> it;
+
+    public ReplayableFirehose(InputRowParser parser) throws IOException
+    {
+      long counter = 0, totalBytes = 0, unparseable = 0, retryCount = 0;
+      Set<String> dimensionScratch = Sets.newHashSet();
+
+      File tmpDir = Files.createTempDir();
+      if (DELETE_TEMP_FILES) {
+        tmpDir.deleteOnExit();
+      }
+
+      long ingestStartMillis = System.currentTimeMillis();
+      do {
+        possiblyDeleteTempFiles();
+        try (Firehose delegateFirehose = delegateFactory.connect(parser)) {
+          while (delegateFirehose.hasMore()) {
+            File tmpFile = File.createTempFile("replayable-", null, tmpDir);
+            if (DELETE_TEMP_FILES) {
+              tmpFile.deleteOnExit();
+            }
+
+            files.add(tmpFile);
+            log.debug("Created file [%s]", tmpFile.getAbsolutePath());
+
+            try (CountingOutputStream cos = new CountingOutputStream(new FileOutputStream(tmpFile));
+                 JsonGenerator generator = factory.createGenerator(cos)) {
+
+              while (delegateFirehose.hasMore() && cos.getCount() < getMaxTempFileSize()) {
+                try {
+                  InputRow row = delegateFirehose.nextRow();
+                  generator.writeObject(row);
+                  dimensionScratch.addAll(row.getDimensions());
+                  counter++;
+                }
+                catch (ParseException e) {
+                  if (reportParseExceptions) {
+                    throw e;
+                  }
+                  unparseable++;
+                }
+              }
+
+              totalBytes += cos.getCount();
+            }
+          }
+          break; // we succeeded so break out of retry loop
+        }
+        catch (Exception e) {
+          if (++retryCount <= readFirehoseRetries && !(e instanceof ParseException)) {
+            log.error(e, "Delegate firehose threw an exception, retrying (%d of %d)", retryCount, readFirehoseRetries);
+          } else {
+            log.error(e, "Delegate firehose threw an exception, retries exhausted, aborting");
+            Throwables.propagate(e);
+          }
+        }
+      } while (true);
+
+      log.info(
+          "Finished reading from firehose in [%,dms], [%,d] events parsed, [%,d] bytes written, [%,d] events unparseable",
+          System.currentTimeMillis() - ingestStartMillis,
+          counter,
+          totalBytes,
+          unparseable
+      );
+
+      dimensions = Ordering.natural().immutableSortedCopy(dimensionScratch);
+
+      if (counter == 0) {
+        log.warn("Firehose contains no events!");
```

Comment:
We can't really print any useful identifier here since it's abstracted by the Firehose interface. It may not even be a file.